File: examples/javascript/test/bidirectional/logInspector.spec.js
const {suite} = require('selenium-webdriver/testing');
const assert = require("assert");
const firefox = require('selenium-webdriver/firefox');
const LogInspector = require('selenium-webdriver/bidi/logInspector');

suite(function (env) {
    describe('Log Inspector', function () {
        let driver

        beforeEach(async function () {
            driver = await env
                .builder()
                .setFirefoxOptions(new firefox.Options().enableBidi())
                .build()
        })

        afterEach(async function () {
            await driver.quit()
        })

        it('test listen to console log', async function () {
            let logEntry = null
            const inspector = await LogInspector(driver)
            await inspector.onConsoleEntry(function (log) {
                logEntry = log
            })

            await driver.get('https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html')
            await driver.findElement({id: 'consoleLog'}).click()

            assert.equal(logEntry.text, 'Hello, world!')
            assert.equal(logEntry.realm, null)
            assert.equal(logEntry.type, 'console')
            assert.equal(logEntry.level, 'info')
            assert.equal(logEntry.method, 'log')
            assert.equal(logEntry.stackTrace, null)
            assert.equal(logEntry.args.length, 1)

            await inspector.close()
        })

        it('test listen to javascript error log', async function () {
            let logEntry = null
            const inspector = await LogInspector(driver)
            await inspector.onJavascriptException(function (log) {
                logEntry = log
            })

            await driver.get('https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html')
            await driver.findElement({id: 'jsException'}).click()

            assert.equal(logEntry.text, 'Error: Not working')
            assert.equal(logEntry.type, 'javascript')
            assert.equal(logEntry.level, 'error')

            await inspector.close()
        })

        it('test retrieve stack trace for a log', async function () {
            let logEntry = null
            const inspector = await LogInspector(driver)
            await inspector.onJavascriptException(function (log) {
                logEntry = log
            })

            await driver.get('https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html')
            await driver.findElement({id: 'jsException'}).click()

            const stackTrace = logEntry.stackTrace
            assert.notEqual(stackTrace, null)
            assert.equal(stackTrace.callFrames.length, 3)

            await inspector.close()
        })

        it('test listen to logs with multiple consumers', async function () {
            let logEntry1 = null
            let logEntry2 = null
            const inspector = await LogInspector(driver)
            await inspector.onJavascriptException(function (log) {
                logEntry1 = log
            })
            await inspector.onJavascriptException(function (log) {
                logEntry2 = log
            })

            await driver.get('https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html')
            await driver.findElement({id: 'jsException'}).click()

            assert.equal(logEntry1.text, 'Error: Not working')
            assert.equal(logEntry1.type, 'javascript')
            assert.equal(logEntry1.level, 'error')

            assert.equal(logEntry2.text, 'Error: Not working')
            assert.equal(logEntry2.type, 'javascript')
            assert.equal(logEntry2.level, 'error')

            await inspector.close()
        })
    })
}, {browsers: ['firefox']})

File: examples/ruby/spec/bidi/log_inspector_spec.rb
# frozen_string_literal: true

require 'selenium-webdriver'

RSpec.describe 'Log Inspector' do
  let(:wait) { Selenium::WebDriver::Wait.new(timeout: 2) }
  let(:driver) do
    options = Selenium::WebDriver::Options.firefox
    options.add_option(:web_socket_url, true)
    Selenium::WebDriver.for :firefox, options: options
  end

  after do
    driver.quit
  end

  it 'test listen to console log' do
    log_entry = nil
    log_inspector = Selenium::WebDriver::BiDi::LogInspector.new(driver)
    log_inspector.on_console_entry { |log| log_entry = log }

    driver.get 'https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html'
    driver.find_element(id: 'consoleLog').click
    wait.until { !log_entry.nil? }

    expect(log_entry).to have_attributes(
      text: 'Hello, world!',
      realm: nil,
      type: 'console',
      level: Selenium::WebDriver::BiDi::LogInspector::LOG_LEVEL[:INFO],
      method: 'log',
      stack_trace: nil
    )
    expect(log_entry.args.size).to eq(1)
  end

  it 'test listen to javascript log' do
    log_entry = nil
    log_inspector = Selenium::WebDriver::BiDi::LogInspector.new(driver)
    log_inspector.on_javascript_log { |log| log_entry = log }

    driver.get 'https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html'
    driver.find_element(id: 'jsException').click
    wait.until { !log_entry.nil? }

    expect(log_entry).to have_attributes(
      text: 'Error: Not working',
      type: 'javascript',
      level: Selenium::WebDriver::BiDi::LogInspector::LOG_LEVEL[:ERROR]
    )
  end

  it 'test listen to javascript error log' do
    log_entry = nil
    log_inspector = Selenium::WebDriver::BiDi::LogInspector.new(driver)
    log_inspector.on_javascript_exception { |log| log_entry = log }

    driver.get 'https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html'
    driver.find_element(id: 'jsException').click
    wait.until { !log_entry.nil? }

    expect(log_entry).to have_attributes(
      text: 'Error: Not working',
      type: 'javascript'
    )
  end

  it 'test retrieve stack trace for a log' do
    log_entry = nil
    log_inspector = Selenium::WebDriver::BiDi::LogInspector.new(driver)
    log_inspector.on_javascript_exception { |log| log_entry = log }

    driver.get 'https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html'
    driver.find_element(id: 'jsException').click
    wait.until { !log_entry.nil? }

    stack_trace = log_entry.stack_trace

    expect(stack_trace).not_to be_nil
    expect(stack_trace['callFrames'].size).to eq(3)
  end

  it 'test listen to logs with multiple consumers' do
    log_entry1 = nil
    log_entry2 = nil
    log_inspector = Selenium::WebDriver::BiDi::LogInspector.new(driver)
    log_inspector.on_javascript_exception { |log| log_entry1 = log }
    log_inspector.on_javascript_exception { |log| log_entry2 = log }

    driver.get 'https://www.selenium.dev/selenium/web/bidi/logEntryAdded.html'
    driver.find_element(id: 'jsException').click
    wait.until { !log_entry1.nil? }
    wait.until { !log_entry2.nil? }

    expect(log_entry1).to have_attributes(
      text: 'Error: Not working',
      type: 'javascript'
    )

    expect(log_entry2).to have_attributes(
      text: 'Error: Not working',
      type: 'javascript'
    )
  end
end
